File: benchmarks/bench_dispatch_scaling.py
#!/usr/bin/env python3
"""PyTorch Dispatch Scaling Benchmark — counterpart to nabla bench_dispatch_scaling.

Measures dispatch overhead across tensor sizes and op counts.
Run: python3 benchmarks/bench_dispatch_scaling.py
"""
import json
import sys
import time
import torch
import torch.nn as nn

WARMUP = 20
SIZES = [32, 64, 128, 256, 512, 1024, 2048, 4096]
CHAIN_OPS = 1000
device = torch.device("cuda")


# ========================================================================
# Experiment 1: Small tensor × many ops (exp chain)
# ========================================================================

def exp1_op_chain():
    print("\n=== Exp1: Small tensor x {} chained exp ops ===".format(CHAIN_OPS), file=sys.stderr)
    print(f"{'size':<8} {'total_ms':>12} {'ops/sec':>14} {'ns/op':>10}", file=sys.stderr)
    print("-" * 48, file=sys.stderr)

    for sz in SIZES:
        a = torch.randn(sz, sz, device=device, dtype=torch.float32)
        for _ in range(WARMUP):
            a = torch.exp(a)
        torch.cuda.synchronize()

        ring = [a.clone() for _ in range(4)]
        torch.cuda.synchronize()
        start = time.perf_counter()
        for i in range(CHAIN_OPS):
            ring[i % 4] = torch.exp(ring[(i + 3) % 4])
        torch.cuda.synchronize()
        elapsed = time.perf_counter() - start

        total_ms = elapsed * 1000
        ops_per_sec = CHAIN_OPS / elapsed
        ns_per_op = elapsed / CHAIN_OPS * 1e9

        print(json.dumps({"test": "exp1_chain", "size": sz, "ops": CHAIN_OPS,
                           "total_ms": round(total_ms, 3),
                           "ops_per_sec": round(ops_per_sec),
                           "ns_per_op": round(ns_per_op)}))
        print(f"{sz:<8} {total_ms:>12.3f} {ops_per_sec:>14.0f} {ns_per_op:>10.0f}", file=sys.stderr)
        del ring


# ========================================================================
# Experiment 2: MLP forward pass (784→256→128→10)
# ========================================================================

class MLP(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc1 = nn.Linear(784, 256)
        self.fc2 = nn.Linear(256, 128)
        self.fc3 = nn.Linear(128, 10)

    def forward(self, x):
        x = torch.relu(self.fc1(x))
        x = torch.relu(self.fc2(x))
        return self.fc3(x)


def exp2_mlp_forward():
    model = MLP().to(device)
    model.eval()
    iters = 500

    print("\n=== Exp2: MLP forward (784→256→128→10) ===", file=sys.stderr)
    print(f"{'batch':<12} {'ms/fwd':>10} {'fwd/sec':>12}", file=sys.stderr)
    print("-" * 36, file=sys.stderr)

    for batch in [1, 32, 128, 1024]:
        x = torch.randn(batch, 784, device=device, dtype=torch.float32)

        with torch.no_grad():
            for _ in range(WARMUP):
                model(x)
            torch.cuda.synchronize()

            start = time.perf_counter()
            for _ in range(iters):
                model(x)
            torch.cuda.synchronize()
            elapsed = time.perf_counter() - start

        ms_per_fwd = elapsed / iters * 1000
        fwd_per_sec = iters / elapsed

        print(json.dumps({"test": "exp2_forward", "batch": batch,
                           "ms_per_fwd": round(ms_per_fwd, 4),
                           "fwd_per_sec": round(fwd_per_sec),
                           "iters": iters}))
        print(f"{batch:<12} {ms_per_fwd:>10.4f} {fwd_per_sec:>12.0f}", file=sys.stderr)


# ========================================================================
# Experiment 3: Training step (forward + backward + SGD step)
# ========================================================================

def exp3_train_step():
    steps = 200

    print("\n=== Exp3: Training step (fwd+bwd+sgd) ===", file=sys.stderr)
    print(f"{'batch':<12} {'ms/step':>12} {'steps/sec':>12}", file=sys.stderr)
    print("-" * 38, file=sys.stderr)

    for batch in [1, 32, 128]:
        model = MLP().to(device)
        model.train()
        optimizer = torch.optim.SGD(model.parameters(), lr=0.01)
        x = torch.randn(batch, 784, device=device, dtype=torch.float32)
        target = torch.randn(batch, 10, device=device, dtype=torch.float32)
        loss_fn = nn.MSELoss()

        # warmup
        for _ in range(5):
            optimizer.zero_grad()
            out = model(x)
            loss = loss_fn(out, target)
            loss.backward()
            optimizer.step()
        torch.cuda.synchronize()

        start = time.perf_counter()
        for _ in range(steps):
            optimizer.zero_grad()
            out = model(x)
            loss = loss_fn(out, target)
            loss.backward()
            optimizer.step()
        torch.cuda.synchronize()
        elapsed = time.perf_counter() - start

        ms_per_step = elapsed / steps * 1000
        steps_per_sec = steps / elapsed

        print(json.dumps({"test": "exp3_train", "batch": batch,
                           "ms_per_step": round(ms_per_step, 4),
                           "steps_per_sec": round(steps_per_sec),
                           "steps": steps}))
        print(f"{batch:<12} {ms_per_step:>12.4f} {steps_per_sec:>12.0f}", file=sys.stderr)


def main():
    print(f"PyTorch {torch.__version__}, CUDA {torch.version.cuda}, {torch.cuda.get_device_name(0)}", file=sys.stderr)
    exp1_op_chain()
    exp2_mlp_forward()
    exp3_train_step()


if __name__ == "__main__":
    main()

File: benchmarks/src/bench_dispatch_scaling.rs
//! Dispatch Scaling Benchmark — dispatch overhead vs tensor size & op count.
//! Run: cargo run --release --bin bench_dispatch_scaling --features cuda

use nabla::prelude::*;
use nabla_bench::{rand_tensor, gpu_sync, kaiming, bench_ms};
use nabla_train::prelude::{Optimizer, Sgd};
use std::time::Instant;

const WARMUP: usize = 20;
const SIZES: [usize; 8] = [32, 64, 128, 256, 512, 1024, 2048, 4096];
const CHAIN_OPS: usize = 1000;

type B = DefaultBackend;

fn exp1_op_chain() {
    eprintln!("\n=== Exp1: Small tensor x {CHAIN_OPS} chained exp ops ===");
    eprintln!("{:<8} {:>12} {:>14} {:>10}", "size", "total_ms", "ops/sec", "ns/op");
    eprintln!("{}", "-".repeat(48));

    for &sz in &SIZES {
        let a = rand_tensor(sz, sz);
        for _ in 0..WARMUP { let _ = a.exp(); }
        gpu_sync();
        let mut ring = vec![a.clone(); 4];
        let start = Instant::now();
        for i in 0..CHAIN_OPS { ring[i % 4] = ring[(i + 3) % 4].exp(); }
        gpu_sync();
        let elapsed = start.elapsed();
        let total_ms = elapsed.as_secs_f64() * 1000.0;
        let ops_per_sec = CHAIN_OPS as f64 / elapsed.as_secs_f64();
        let ns_per_op = elapsed.as_nanos() as f64 / CHAIN_OPS as f64;
        println!("{{\"test\":\"exp1_chain\",\"size\":{sz},\"ops\":{CHAIN_OPS},\"total_ms\":{total_ms:.3},\"ops_per_sec\":{ops_per_sec:.0},\"ns_per_op\":{ns_per_op:.0}}}");
        eprintln!("{sz:<8} {total_ms:>12.3} {ops_per_sec:>14.0} {ns_per_op:>10.0}");
        drop(ring);
    }
}

fn exp2_mlp_forward() {
    let w1 = kaiming(784, 256);
    let w2 = kaiming(256, 128);
    let w3 = kaiming(128, 10);

    eprintln!("\n=== Exp2: MLP forward (784→256→128→10) ===");
    eprintln!("{:<12} {:>10} {:>12}", "batch", "ms/fwd", "fwd/sec");
    eprintln!("{}", "-".repeat(36));

    for &(batch, iters) in &[(1usize, 500usize), (32, 100), (128, 50), (1024, 20)] {
        let x = rand_tensor(batch, 784);
        let fwd = |x: &Tensor<f32>| {
            let h1 = (x * &w1).leaky_relu(0.01);
            let h2 = (&h1 * &w2).leaky_relu(0.01);
            &h2 * &w3
        };
        for _ in 0..10 { let _ = fwd(&x); }
        gpu_sync();
        let mut ring = vec![fwd(&x); 4];
        let start = Instant::now();
        for i in 0..iters { ring[i % 4] = fwd(&x); }
        gpu_sync();
        let elapsed = start.elapsed();
        drop(ring);
        let ms_per_fwd = elapsed.as_secs_f64() * 1000.0 / iters as f64;
        let fwd_per_sec = iters as f64 / elapsed.as_secs_f64();
        println!("{{\"test\":\"exp2_forward\",\"batch\":{batch},\"ms_per_fwd\":{ms_per_fwd:.4},\"fwd_per_sec\":{fwd_per_sec:.0},\"iters\":{iters}}}");
        eprintln!("{batch:<12} {ms_per_fwd:>10.4} {fwd_per_sec:>12.0}");
    }
}

fn exp3_train_step() {
    eprintln!("\n=== Exp3: Training step (fwd+bwd+sgd) ===");
    eprintln!("{:<12} {:>12} {:>12}", "batch", "ms/step", "steps/sec");
    eprintln!("{}", "-".repeat(38));

    for &(batch, steps) in &[(1usize, 100usize), (32, 30), (128, 10)] {
        let mut w1 = kaiming(784, 256);
        let mut w2 = kaiming(256, 128);
        let mut w3 = kaiming(128, 10);
        let params = vec![&w1, &w2, &w3];
        let mut sgd: Sgd<f32, B> = Sgd::from_params(0.01, &params);
        drop(params);
        let x = rand_tensor(batch, 784);
        let target = rand_tensor(batch, 10);

        let mut do_step = |w1: &mut Tensor<f32>, w2: &mut Tensor<f32>, w3: &mut Tensor<f32>| {
            let tape = Tape::new();
            if let (Ok(x_var), Ok(t_var)) = (tape.variable(x.clone()), tape.variable(target.clone())) {
                let w1v = tape.variable(w1.clone()).ok();
                let w2v = tape.variable(w2.clone()).ok();
                let w3v = tape.variable(w3.clone()).ok();
                if let (Some(w1v), Some(w2v), Some(w3v)) = (w1v, w2v, w3v) {
                    let h1 = x_var.matmul(&w1v).leaky_relu(0.01);
                    let h2 = h1.matmul(&w2v).leaky_relu(0.01);
                    let out = h2.matmul(&w3v);
                    let diff = out.sub_var(&t_var);
                    let loss = diff.emul(&diff).sum_axis(1).sum_axis(0);
                    if loss.backward().is_ok() {
                        let g0 = w1v.grad_ref().ok();
                        let g1 = w2v.grad_ref().ok();
                        let g2 = w3v.grad_ref().ok();
                        if let (Some(r0), Some(r1), Some(r2)) = (&g0, &g1, &g2) {
                            sgd.step(&mut vec![w1, w2, w3], &[&**r0, &**r1, &**r2]);
                        }
                    }
                }
            }
        };

        for _ in 0..3 { do_step(&mut w1, &mut w2, &mut w3); }
        gpu_sync();

        let ms = bench_ms(0, steps, || do_step(&mut w1, &mut w2, &mut w3));
        let steps_per_sec = 1000.0 / ms;
        println!("{{\"test\":\"exp3_train\",\"batch\":{batch},\"ms_per_step\":{ms:.4},\"steps_per_sec\":{steps_per_sec:.0},\"steps\":{steps}}}");
        eprintln!("{batch:<12} {ms:>12.4} {steps_per_sec:>12.0}");
    }
}

fn main() {
    exp1_op_chain();
    exp2_mlp_forward();
    exp3_train_step();
}
